TypeORM tests fail on MySQL 8.0

## Issue Description

Tracking issue.

### Expected Behavior

Tests pass on MySQL 8.0.


### Actual Behavior

Tests fail with:

```

2) database schema > column width
   all types should be created with correct width:
 AssertionError: expected undefined to equal 10
  at /typeorm/test/functional/database-schema/column-width/mysql/column-width.ts:25:61

3) database schema > column width
   should update data type display width:
 AssertionError: expected undefined to equal 5
  at /typeorm/test/functional/database-schema/column-width/mysql/column-width.ts:48:61

4) entity-schema > columns > mysql
   should create columns with different options:

  AssertionError: expected 'concat(`FirstName`,_utf8mb4\\\' \\\',`LastName`)' to equal 'concat(`FirstName`,\' \',`LastName`)'
  + expected - actual

  -concat(`FirstName`,_utf8mb4\' \',`LastName`)
  +concat(`FirstName`,' ',`LastName`)
  
  at /typeorm/test/functional/entity-schema/columns/mysql/columns-mysql.ts:26:77

5) transaction > method wrapped into transaction decorator
   should execute all operations in the method in a transaction:
 AssertionError: expected undefined not to be undefined
  at /typeorm/test/functional/transaction/transaction-decorator/transaction-decorator.ts:35:37

6) transaction > method wrapped into transaction decorator
   should inject repository and custom repository into method decorated with @Transaction:
 AssertionError: expected undefined not to be undefined
  at /typeorm/test/functional/transaction/transaction-decorator/transaction-decorator.ts:135:37

7) transaction > method wrapped into transaction decorator
   should execute all operations in the method in a transaction with a specified isolation:
 AssertionError: expected undefined not to be undefined
  at /typeorm/test/functional/transaction/transaction-decorator/transaction-decorator.ts:156:37

8) github issues > #1377 Add support for `GENERATED ALWAYS AS` in MySQL
   should correctly create table with generated columns:

  AssertionError: expected 'concat(`firstName`,_utf8mb4\\\' \\\',`lastName`)' to equal 'concat(`firstName`,\' \',`lastName`)'
  + expected - actual

  -concat(`firstName`,_utf8mb4\' \',`lastName`)
  +concat(`firstName`,' ',`lastName`)
  
  at /typeorm/test/github-issues/1377/issue-1377.ts:22:77

9) github issues > #3702 MySQL Spatial Type Support : GeomFromText function is not supported
   when legacySpatialSupport: true
     should use GeomFromText:
 AssertionError: expected 'INSERT INTO `letter_box`(`id`, `coord`) VALUES (DEFAULT, ST_GeomFromText(?, 4326))' to not include 'ST_GeomFromText'
  at /typeorm/test/github-issues/3702/issue-3702.ts:31:29

10) github issues > #3702 MySQL Spatial Type Support : GeomFromText function is not supported
   when legacySpatialSupport: true
     should use AsText:
 AssertionError: expected 'SELECT `letterBox`.`id` AS `letterBox_id`, ST_AsText(`letterBox`.`coord`) AS `letterBox_coord` FROM `letter_box` `letterBox`' to not include 'ST_AsText'
  at /typeorm/test/github-issues/3702/issue-3702.ts:53:29

11) github issues > #6266 Many identical selects after insert bunch of items
   should execute a single SELECT to get inserted default and generated values of multiple entities:
 TypeError: Attempted to wrap select which is already wrapped
  at checkWrappedMethod (node_modules/sinon/lib/sinon/util/core/wrap-method.js:40:21)
  at wrapMethod (node_modules/sinon/lib/sinon/util/core/wrap-method.js:88:13)
  at Function.spy (node_modules/sinon/lib/sinon/spy.js:156:16)
  at Sandbox.spy (node_modules/sinon/lib/sinon/sandbox.js:328:35)
  at /typeorm/test/github-issues/6266/issue-6266.ts:43:41
  at step (node_modules/tslib/tslib.js:143:27)
  at Object.next (node_modules/tslib/tslib.js:124:57)
  at /typeorm/node_modules/tslib/tslib.js:117:75
  at new Promise (<anonymous>)
  at __awaiter (node_modules/tslib/tslib.js:113:16)
  at /typeorm/test/github-issues/6266/issue-6266.ts:42:46
  at Array.map (<anonymous>)
  at Context.<anonymous> (test/github-issues/6266/issue-6266.ts:42:25)
  at runMicrotasks (<anonymous>)
  at processTicksAndRejections (node:internal/process/task_queues:96:5)
--------------
Error: Stack Trace for original
  at wrapMethod (node_modules/sinon/lib/sinon/util/core/wrap-method.js:116:26)
  at Function.spy (node_modules/sinon/lib/sinon/spy.js:156:16)
  at Sandbox.spy (node_modules/sinon/lib/sinon/sandbox.js:328:35)
  at /typeorm/test/github-issues/6266/issue-6266.ts:43:41
  at step (node_modules/tslib/tslib.js:143:27)
  at Object.next (node_modules/tslib/tslib.js:124:57)
  at /typeorm/node_modules/tslib/tslib.js:117:75
  at new Promise (<anonymous>)
  at __awaiter (node_modules/tslib/tslib.js:113:16)
  at /typeorm/test/github-issues/6266/issue-6266.ts:42:46
  at Array.map (<anonymous>)
  at Context.<anonymous> (test/github-issues/6266/issue-6266.ts:42:25)
  at runMicrotasks (<anonymous>)
  at processTicksAndRejections (node:internal/process/task_queues:96:5)
```

### My Environment

<!--
  Please add any other relevant dependencies to this table at the end.
  For example: Electron, React Native, or NestJS.
-->

| Dependency          | Version  |
| ---                 | ---      |
| Operating System    |          |
| Node.js version     | x.y.zzz  | <!-- run `node -v` to obtain this -->
| Typescript version  | x.y.zzz  | <!-- run `npm list typescript` to obtain this -->
| TypeORM version     | x.y.zzz  | <!-- run `npm list typeorm` to obtain this -->


### Additional Context

<!--
  Add any other context about the bug report here.
-->


### Relevant Database Driver(s)

<!-- Put "yes" instead of "no" to your issue's relevant databases -->

| DB Type              | Reproducible  |
| ---                  | ---           |
| `aurora-data-api`    | no            |
| `aurora-data-api-pg` | no            |
| `better-sqlite3`     | no            |
| `cockroachdb`        | no            |
| `cordova`            | no            |
| `expo`               | no            |
| `mongodb`            | no            |
| `mysql`              | yes            |
| `nativescript`       | no            |
| `oracle`             | no            |
| `postgres`           | no            |
| `react-native`       | no            |
| `sap`                | no            |
| `sqlite`             | no            |
| `sqlite-abstract`    | no            |
| `sqljs`              | no            |
| `sqlserver`          | no            |


### Are you willing to resolve this issue by submitting a Pull Request?

<!-- Put "✅" (:white_check_mark:) to one of these options, left "✖️" (:heavy_multiplication_x:) others: -->

 - ✖️ Yes, I have the time, and I know how to start.
 - ✅ Yes, I have the time, but I don't know how to start. I would need guidance.
 - ✖️ No, I don’t have the time, but I can support (using donations) development.
 - ✖️ No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

--- a/src/driver/mysql/MysqlConnectionOptions.ts
+++ b/src/driver/mysql/MysqlConnectionOptions.ts
@@ -103,6 +103,12 @@ export interface MysqlConnectionOptions
     readonly flags?: string[]
 
     /**
+     * TypeORM will automatically use package found in your node_modules, prioritizing mysql over mysql2,
+     * but you can specify it manually
+     */
+    readonly connectorPackage?: "mysql" | "mysql2"
+
+    /**
      * Replication setup.
      */
     readonly replication?: {

--- a/src/driver/mysql/MysqlDriver.ts
+++ b/src/driver/mysql/MysqlDriver.ts
@@ -1057,9 +1057,15 @@ export class MysqlDriver implements Driver {
      * Loads all driver dependencies.
      */
     protected loadDependencies(): void {
+        const connectorPackage = this.options.connectorPackage ?? "mysql"
+        const fallbackConnectorPackage =
+            connectorPackage === "mysql"
+                ? ("mysql2" as const)
+                : ("mysql" as const)
         try {
             // try to load first supported package
-            const mysql = this.options.driver || PlatformTools.load("mysql")
+            const mysql =
+                this.options.driver || PlatformTools.load(connectorPackage)
             this.mysql = mysql
             /*
              * Some frameworks (such as Jest) may mess up Node's require cache and provide garbage for the 'mysql' module
@@ -1070,14 +1076,17 @@ export class MysqlDriver implements Driver {
              */
             if (Object.keys(this.mysql).length === 0) {
                 throw new TypeORMError(
-                    "'mysql' was found but it is empty. Falling back to 'mysql2'.",
+                    `'${connectorPackage}' was found but it is empty. Falling back to '${fallbackConnectorPackage}'.`,
                 )
             }
         } catch (e) {
             try {
-                this.mysql = PlatformTools.load("mysql2") // try to load second supported package
+                this.mysql = PlatformTools.load(fallbackConnectorPackage) // try to load second supported package
             } catch (e) {
-                throw new DriverPackageNotInstalledError("Mysql", "mysql")
+                throw new DriverPackageNotInstalledError(
+                    "Mysql",
+                    connectorPackage,
+                )
             }
         }
     }

